uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php

```diff
@@ -138,6 +147,232 @@ private static function validateMandatory($params) {
+    $todayDate = new DateTime('today');
+    $leaveDatesHasPastDates = $fromDate < $todayDate || $toDate < $todayDate;
+
+    if ($leaveDatesHasPastDates && !$absenceType->allow_accrue_in_the_past) {
+      throw new InvalidLeaveRequestException(
+        'You cannot request TOIL for past days',
+        'leave_request_toil_cannot_be_requested_for_past_days',
+        'from_date'
+      );
+    }
+  }
+
+  /**
+   * Validate that the TOIL amount to be accrued plus total approved accrued TOIL for the period
+   * is not greater than the maximum defined(if any) for the Absence Type
+   *
+   * @param array $params
+   *   The params array received by the create method
+   *
+   * @throws \CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   */
+  private static function validateTOILToAccruedAmountIsValid($params) {
+    $absenceType = AbsenceType::findById($params['type_id']);
+    $unlimitedAccrual = empty($absenceType->max_leave_accrual) && $absenceType->max_leave_accrual != 0;
+
+    $periodContainingToilDates = AbsencePeriod::getPeriodContainingDates(
+      new DateTime($params['from_date']),
+      new DateTime($params['to_date'])
+    );
+
+    $totalApprovedToilForPeriod = self::getTotalApprovedToilForPeriod(
+      $periodContainingToilDates,
+      $params['contact_id'],
+      $params['type_id']
+    );
+    $totalProjectedToilForPeriod = $totalApprovedToilForPeriod + $params['toil_to_accrue'];
+
+    if ($totalProjectedToilForPeriod > $absenceType->max_leave_accrual && !$unlimitedAccrual) {
+      throw new InvalidLeaveRequestException(
+        'The TOIL amount plus all approved TOIL for current period is greater than the maximum for this Absence Type',
+        'leave_request_toil_amount_more_than_maximum_for_absence_type',
+        'toil_to_accrue'
+      );
+    }
+  }
+
+
+  /**
+   * This method returns the total sum of TOIL accrued for an absence type by a contact
+   * over a given absence period
+   *
+   * @param int $contactID
+   * @param int $typeID
+   * @param \CRM_HRLeaveAndAbsences_BAO_AbsencePeriod $period
+   *
+   * @return float
+   */
+  private static function getTotalApprovedToilForPeriod(AbsencePeriod $period, $contactID, $typeID) {
+    $leaveRequestStatuses = array_flip(self::buildOptions('status_id'));
+
+    $leaveRequestStatusFilter = [
+      $leaveRequestStatuses['Approved'],
+      $leaveRequestStatuses['Admin Approved']
+    ];
+
+    $totalApprovedTOIL = LeaveBalanceChange::getTotalTOILBalanceChangeForContact(
+      $contactID,
+      $typeID,
+      new DateTime($period->start_date),
+      new DateTime($period->end_date),
+      $leaveRequestStatusFilter
+    );
+
+    return $totalApprovedTOIL;
+  }
+
+  /**
+   * Validates the Sickness mandatory fields according to the value of request_type.
+   *
+   * @param array $params
+   *
+   * @throws \CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   */
+  private static function validateSicknessMandatoryFields($params) {
```

Comment:
This method also validates that are supposed to be empty when the request type is not sickness so I think we need a more suitable name.